openedx/openedx-platform
# [BD-04] Convert Randomize XModule to XBlock

Converts the Randomize XModule into an XBlock.

Part of [XModule to XBlock Conversion work](https://openedx.atlassian.net/wiki/spaces/AC/pages/1472790755/XModule+to+XBlock+Conversion). 

**Testing instructions**:

1. It is not possible to create Randomize components from the Studio UI. So create a course and import the attached course archive: [course-randomize.tar.gz](https://github.com/edx/edx-platform/files/5115618/course-randomize.tar.gz)
2. Check that the XModule works. The decision of which child to show is based on the `User.pk` field of the current user. So to see different problems, try viewing the Randomize XModule from different users.
3. Make an export of the course.
4. Update devstack with this code.
   - Cherry-pick this commit.
   - In both `make lms-shell` and `make studio-shell`, install the updated version of the `xmodule` package with `pip install -e /common/lib/xmodule`.
   - Do `make lms-restart` and `make studio-restart`.
5. Test the existing components in both Studio and LMS.
7. Import the exported archive into a new course and check that it still works the same.


**MRO Analysis**
Descriptor MRO | Module MRO | Block MRO | Notes
------------ | ------------- | ---------------- | ------------
<class 'xmodule.randomize_module.RandomizeDescriptor'> | <class 'xmodule.randomize_module.RandomizeModule'> | <class 'xmodule.randomize_module.RandomizeBlock'> | Merged into one class.
<class 'xmodule.randomize_module.RandomizeFields'> | <class 'xmodule.randomize_module.RandomizeFields'> | — | Merged in RandomizeBlock class
<class 'xmodule.seq_module.SequenceDescriptor'> | — | — | Attributes needed by RandomizeBlock have been split out into SequenceMixin
<class 'xmodule.seq_module.SequenceMixin'> | — | <class 'xmodule.seq_module.SequenceMixin'> | —
<class 'xmodule.seq_module.SequenceFields'> | — | <class 'xmodule.seq_module.SequenceFields'>
<class 'xmodule.seq_module.ProctoringFields'> | — | — | Not needed by RandomizeBlock
<class 'xmodule.mako_module.MakoModuleDescriptor'> | — | — | Inherits from (MakoTemplateBlockBase, XModuleDescriptor) and has the get_html() method which has been implemented in the XBlock.
<class 'xmodule.mako_module.MakoTemplateBlockBase'> | — | <class 'xmodule.mako_module.MakoTemplateBlockBase'> |
<class 'xmodule.xml_module.XmlDescriptor'> | — | — | Inherits from (XmlMixin, XModuleDescriptor) and doens't have any atttributes so we can remove it.
<class 'xmodule.xml_module.XmlMixin'> | — | <class 'xmodule.xml_module.XmlMixin'> | —
<class 'xmodule.xml_module.XmlParserMixin'> | — | <class 'xmodule.xml_module.XmlParserMixin'> | —
<class 'xmodule.x_module.XModuleDescriptor'> | — | — | Not needed.
<class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | — | <class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | —
— | <class 'xmodule.x_module.XModule'> | — | Not needed.
— | <class 'xmodule.x_module.XModuleToXBlockMixin'> | <class 'xmodule.x_module.XModuleToXBlockMixin'> | —
<class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | —
<class 'xmodule.x_module.ResourceTemplates'> | — | <class 'xmodule.x_module.ResourceTemplates'> | —
<class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | —
<class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | —
<class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | —
